pyrival/graphs/maximum_matching.py

```diff
@@ -0,0 +1,66 @@
+import itertools
+from random import randint
+
+
+_DEFAULT_PRIME = 1073750017
+
+
+def maximum_matching(edges, mod=_DEFAULT_PRIME):
+    """
+    Returns the maximum cardinality matching of an undirected general graph
+    Uses a randomized algorithm to compute the rank of the Tutte matrix
+    The rank of the Tutte matrix is equal to twice the size of the maximum matching with high probability
+
+    Complexity: O(n ^ 3) worst case, O(n * m) with high constant factor on average
+
+    :param edges: a list of edges, assume nodes can be anything numbered from 0 to max number in edges
+    :param mod: a large random prime
+    :return: the maximum cardinality matching of the graph
+    """
+
+    edges = [(u, v) for u, v in edges if u != v]
+
+    if not edges:
+        return 0
+
+    n = max(itertools.chain(*edges)) + 1
+    matrix = _get_tutte_matrix(n, edges, mod)
+    return _gauss(matrix, mod) // 2
+
+
+def _get_tutte_matrix(n, edges, mod):
+    matrix = [[0] * n for _ in range(n)]
+
+    for u, v in edges:
+        u, v = min(u, v), max(u, v)
+        val = randint(1, mod - 1)
+        matrix[u][v], matrix[v][u] = val, mod - val
+
+    return matrix
+
+
+def _gauss(matrix, mod):
+    r, n = 0, len(matrix)
+    for j in range(n):
+        k = r
+        while k < n and not matrix[k][j]:
+            k += 1
+
+        if k == n:
+            continue
+
+        inv = pow(matrix[k][j], mod - 2, mod)
+        for i in range(n):
+            matrix[k][i], matrix[r][i] = matrix[r][i], inv * matrix[k][i] % mod
+
+        for u in range(r + 1, n):
+            if not matrix[u][j]:
+                continue
+
+            for v in range(j + 1, n):
+                if matrix[r][v]:
+                    matrix[u][v] = (matrix[u][v] - matrix[r][v] * matrix[u][j]) % mod
```

Comment:
@bjorn-martinsson it is indeed partially translated from my old C implementation :-)

But that is not the reason I initially included it. I was unaware Python's modulo operator behaves differently. Learnt a new thing today. I always assumed it was similar to C/C++ without explicitly checking for myself. Removed the redundant check.

Reply:
The first one (`if not matrix[u][j]:`) is quite useful because it won't enter the inner loop at all. In sparse graphs this is quite useful and will result in the algorithm being several times faster. The second one is useful because of the reason you explained above, `long` multiplication followed by expensive modulo is slow. Here are the `cProfile` stats if I run the method with local cPython for the large test case (`n=500, m=6500)`.

Without these `if` checks:
```
   39328 function calls in 8.170 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.002    0.002    8.170    8.170 <string>:1(<module>)
        1    0.001    0.001    0.001    0.001 lol.py:21(<listcomp>)
        1    0.002    0.002    0.012    0.012 lol.py:31(_get_tutte_matrix)
        1    0.001    0.001    0.001    0.001 lol.py:32(<listcomp>)
        1    8.154    8.154    8.155    8.155 lol.py:41(_gauss)
        1    0.000    0.000    8.168    8.168 lol.py:8(maximum_matching)
     6490    0.003    0.000    0.007    0.000 random.py:170(randrange)
     6490    0.002    0.000    0.009    0.000 random.py:214(randint)
     6490    0.003    0.000    0.004    0.000 random.py:220(_randbelow)
        1    0.000    0.000    8.170    8.170 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.max}
      478    0.002    0.000    0.002    0.000 {built-in method builtins.pow}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.print}
     6490    0.000    0.000    0.000    0.000 {method 'bit_length' of 'int' objects}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
    12879    0.001    0.000    0.001    0.000 {method 'getrandbits' of '_random.Random' objects}
```

With `if matrix[r][v]:` only:
```
   39526 function calls in 6.078 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.002    0.002    6.078    6.078 <string>:1(<module>)
        1    0.001    0.001    0.001    0.001 lol.py:21(<listcomp>)
        1    0.002    0.002    0.012    0.012 lol.py:31(_get_tutte_matrix)
        1    0.001    0.001    0.001    0.001 lol.py:32(<listcomp>)
        1    6.061    6.061    6.063    6.063 lol.py:41(_gauss)
        1    0.000    0.000    6.076    6.076 lol.py:8(maximum_matching)
     6490    0.003    0.000    0.008    0.000 random.py:170(randrange)
     6490    0.001    0.000    0.009    0.000 random.py:214(randint)
     6490    0.003    0.000    0.005    0.000 random.py:220(_randbelow)
        1    0.000    0.000    6.078    6.078 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.max}
      478    0.002    0.000    0.002    0.000 {built-in method builtins.pow}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.print}
     6490    0.000    0.000    0.000    0.000 {method 'bit_length' of 'int' objects}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
    13077    0.001    0.000    0.001    0.000 {method 'getrandbits' of '_random.Random' objects}
```

With both checks:
```
   39465 function calls in 3.453 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.002    0.002    3.452    3.452 <string>:1(<module>)
        1    0.001    0.001    0.001    0.001 lol.py:21(<listcomp>)
        1    0.002    0.002    0.012    0.012 lol.py:31(_get_tutte_matrix)
        1    0.001    0.001    0.001    0.001 lol.py:32(<listcomp>)
        1    3.436    3.436    3.437    3.437 lol.py:41(_gauss)
        1    0.000    0.000    3.451    3.451 lol.py:8(maximum_matching)
     6490    0.003    0.000    0.008    0.000 random.py:170(randrange)
     6490    0.002    0.000    0.009    0.000 random.py:214(randint)
     6490    0.003    0.000    0.005    0.000 random.py:220(_randbelow)
        1    0.000    0.000    3.453    3.453 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.max}
      478    0.002    0.000    0.002    0.000 {built-in method builtins.pow}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.print}
     6490    0.000    0.000    0.000    0.000 {method 'bit_length' of 'int' objects}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
    13016    0.001    0.000    0.001    0.000 {method 'getrandbits' of '_random.Random' objects}
```